blockchain/store_test.go

```diff
@@ -0,0 +1,263 @@
+package blockchain_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"runtime"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/tendermint/go-wire/data"
+	"github.com/tendermint/tendermint/blockchain"
+	"github.com/tendermint/tendermint/types"
+	"github.com/tendermint/tmlibs/db"
+)
+
+func TestLoadBlockStoreStateJSON(t *testing.T) {
+	db := db.NewMemDB()
+
+	bsj := &blockchain.BlockStoreStateJSON{Height: 1000}
+	bsj.Save(db)
+
+	retrBSJ := blockchain.LoadBlockStoreStateJSON(db)
+
+	assert.Equal(t, *bsj, retrBSJ, "expected the retrieved DBs to match")
+}
+
+var blockStoreKey = []byte("blockStore")
+
+func TestNewBlockStore(t *testing.T) {
+	db := db.NewMemDB()
+	db.Set(blockStoreKey, []byte(`{"height": 10000}`))
+	bs := blockchain.NewBlockStore(db)
+	assert.Equal(t, bs.Height(), 10000, "failed to properly parse blockstore")
+
+	panicCausers := [][]byte{
+		[]byte("artful-doger"),
+		[]byte(""),
+		[]byte(" "),
+	}
+
+	for i, panicCauser := range panicCausers {
+		// Expecting a panic here on trying to parse an invalid blockStore
+		_, _, err := doFn(func() (interface{}, error) {
+			db.Set(blockStoreKey, panicCauser)
+			_ = blockchain.NewBlockStore(db)
+			return nil, nil
+		})
+		assert.NotNil(t, err, "#%d panicCauser: %q expected a panic", i, panicCauser)
```

Comment:
there is an https://godoc.org/github.com/stretchr/testify/assert#Panics

Reply:
Nice, thank you for the reference, I hadn't seen that function. I'll skip using it though for this test particularly because for some of these we want panics, for others we don't and constructing the code with assert.Panics might be a little hairy, I haven't seen an easy way of doing it :)

Reply:
arent all three of these test cases expecting a panic ?

Reply:
@ebuchman in this isolated test with these 3 test cases, yes all those 3 expect panics, but doFn is also used throughout the file in other cases that don't necessarily want to panic.